Lib9c/Action/MigrationLegacyShop.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Bencodex.Types;
+using Libplanet;
+using Libplanet.Action;
+using Nekoyume.Model.State;
+using Serilog;
+
+namespace Nekoyume.Action
+{
+    [Serializable]
+    [ActionType("migration_legacy_shop")]
+    public class MigrationLegacyShop : GameAction
+    {
+        public List<Address> SellerAvatarAddresses;
+        public override IAccountStateDelta Execute(IActionContext context)
+        {
+            var states = context.PreviousStates;
+            if (context.Rehearsal)
+            {
+                states = SellerAvatarAddresses.Aggregate(states,
+                    (current, sellerAvatarAddress) => current.SetState(sellerAvatarAddress, MarkChanged));
+
+                return states.SetState(Addresses.Shop, MarkChanged);
+            }
+
+            CheckPermission(context);
+
+            Log.Debug("Start Migration Legacy Shop");
+            var shopState = states.GetShopState();
+            var groupBy = shopState.Products.Values
+                .Where(s => SellerAvatarAddresses.Contains(s.SellerAvatarAddress))
+                .GroupBy(p => p.SellerAvatarAddress)
+                .ToList();
+
+            Log.Debug($"Found {groupBy.Count} Seller Avatar.");
+            foreach (var group in groupBy)
+            {
+                var avatarAddress = group.Key;
+                var avatarState = states.GetAvatarState(avatarAddress);
```

Comment:
how about caching `avatarState` for the same `avatarAddress`?

Reply:
i try caching `avatarState`, but cache does not work. because `GetState` is called only once per avatar address.